octree/include/pcl/octree/octree_iterator.h

```diff
@@ -524,6 +523,77 @@ namespace pcl
       };
 
     //////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+    /** \brief @b Octree iterator class
+     * \note This class implements a forward iterator for traversing octrees in a breadth-first manner.
+     *       It only returns the node at a given depth in the octree, if any.
+     * \ingroup octree
+     * \author Fabien Rozar (fabien.rozar@gmail.com)
+     */
+    template<typename OctreeT>
+      class OctreeBreadthFirstAtDepthIterator : public OctreeBreadthFirstIterator<OctreeT>
+      {
+      public:
+
+        // public typedefs
+        typedef typename OctreeBreadthFirstIterator<OctreeT>::BranchNode BranchNode;
+        typedef typename OctreeBreadthFirstIterator<OctreeT>::LeafNode LeafNode;
+
+        /** \brief Empty constructor.
+          * \param[in] depth_arg Depth level during traversal
+          */
+        explicit
+        OctreeBreadthFirstAtDepthIterator (unsigned int depth_arg = 0);
+
+        /** \brief Constructor.
+        * \param[in] octree_arg Octree to be iterated. Initially the iterator is set to its root node.
+        * \param[in] depth_arg Depth level during traversal
+        */
+        explicit
+        OctreeBreadthFirstAtDepthIterator (OctreeT* octree_arg, unsigned int depth_arg = 0);
+
+        /** \brief Empty deconstructor. */
+        virtual
+        ~OctreeBreadthFirstAtDepthIterator ();
+
+        /** \brief Copy operator.
+        * \param[in] src the iterator to copy into this
+        */
+        inline OctreeBreadthFirstAtDepthIterator&
+        operator = (const OctreeBreadthFirstAtDepthIterator& src)
+        {
+          OctreeBreadthFirstIterator<OctreeT>::operator=(src);
```

Comment:
Missing space

Reply:
If you're defining copy assignment, you should define as well the copy constructor.

Reply:
Assuming your member depth_ will be const, what happens if you're copying from an iterator with a different depth? Assuming your depth_ won't be const, you should also copy it.

Reply:
As you expected, the member `depth_` cannot be a const member because of the `operator =` (I get the compilation error: assignment of read-only member).

Reply:
A copy constructor is also implemented.